update_copyright.d

```diff
@@ -0,0 +1,371 @@
+#!/usr/bin/env rdmd
+
+/**
+Update the copyright notices in source files so that they have the form:
+---
+    Copyright XXXX-YYYY by The D Language Foundation, All Rights Reserved
+---
+It does not change copyright notices of authors that are known to have made
+changes under a proprietary license.
+
+Copyright:  Copyright (C) 2017-2018 by The D Language Foundation, All Rights Reserved
+
+License:    $(WEB boost.org/LICENSE_1_0.txt, Boost License 1.0)
+
+Authors:    Iain Buclaw
+
+Example usage:
+---
+update-copyright.py --update-year src/dmd
+---
+*/
+
+module tools.update_copyright;
+
+int main(string[] args)
+{
+    import std.getopt;
+
+    bool updateYear;
+    bool verbose;
+    auto opts = getopt(args,
+        "update-year|y", "Update the current year on every notice", &updateYear,
+        "verbose|v", "Be more verbose", &verbose);
+
+    if (args.length == 1 || opts.helpWanted)
+    {
+        defaultGetoptPrinter("usage: update_copyright [--help|-h] [--update-year|-y] <dir>...",
+                             opts.options);
+        return 0;
+    }
+
+    Copyright(updateYear, verbose).run(args[1 .. $]);
+    return 0;
+}
+
+//::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
+
+struct Copyright
+{
+    import std.algorithm, std.array, std.file, std.regex, std.stdio, std.string;
+
+    // The author to use in copyright notices.
+    enum author = "The D Language Foundation, All Rights Reserved";
+
+    // The standard (C) form.
+    enum copyright = "(C)";
+
+private:
+    // True if running in verbose mode.
+    bool verbose = false;
+
+    // True if also updating copyright year.
+    bool updateYear = false;
+
+    // An associative array of known copyright holders.
+    // Value set is true if the copyright holder is internal.
+    bool[string] holders;
+
+    // Files and directories to ignore during search.
+    static string[] skipDirs = [
+        "docs",
+        "ini",
+        "test",
+        "samples",
+        "vcbuild",
+        ".git",
+    ];
+
+    static string[] skipFiles = [
+        "Jenkinsfile",
+        "LICENSE.txt",
+        "VERSION",
+        ".a",
+        ".ddoc",
+        ".deps",
+        ".lst",
+        ".map",
+        ".md",
+        ".o",
+        ".obj",
+        ".sdl",
+        ".sh",
+        ".yml",
+    ];
+
+    // Characters in a range of years.
+    // Include '.' for typos, and '?' for unknown years.
+    enum rangesStr = `[0-9?](?:[-0-9.,\s]|\s+and\s+)*[0-9]`;
+
+    // Non-whitespace characters in a copyright holder's name.
+    enum nameStr = `[\w.,-]`;
+
+    // Matches a full copyright notice:
+    // - 'Copyright (C)', etc.
+    // - The years. Includes the whitespace in the year, so that we can
+    //   remove any excess.
+    // - 'by ', if used
+    // - The copyright holder.
+    Regex!char copyrightRe;
```

Comment:
Might be nice to assign these regexes here where they are being declared.  They can also be static right? (Not that it's a big deal for such a small file...but still).

Reply:
`ctRegex` adds dozens of seconds onto the compile-time.

Reply:
I didn't mean to use `ctRegex`.  I just mean to assign them at their declaration.  Also I don't think addressed my comment about changing them to static.

Reply:
Marking them static just increases binary size.

Reply:
Can you provide the numbers?

Reply:
Out of curiosity I went ahead and checked, and it looks like it does increase the binary size by about 100K.  I didn't know regex'es took so much space.  I've also never seen someone use this technique to decrease the binary size...interesting stuff.  This is sortof akin to "compression", you leave the regex in string form, and keep the code that "decompresses" it into it's equivalent data structures (which turn out to be large).  I was thinking that if you made them static, the compiler may be able to optimize out the code that processes a regex, but it doesn't appear that way (even when I enabled optimization).  Anyway, seems fine to leave them as instance members.   Though I might still consider moving their initialization to where they are declared.  Currently the 2 are so far away... :)

Reply:
Also making them `const` might be nice...

Reply:
Ha, that makes a brave assumption that `std.regex` is const-safe. :-)